### PR TITLE
CSS-6244 Adds client ID and secret to the provider config.

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

Added client ID and client secret to Juju provider config, which will allow terraform provider to log in to JAAS using these client credentials.

Updates Juju dependency to 3.5 that contains needed login changes.

## Type of change

*\<What type of a change is this? Please keep one or more relevant lines from below and delete the others.\>*

- Other: login method change

## Environment

- Juju controller version: 3.5

- Terraform version: 

## QA steps

Manual QA steps should be done to test this PR.

```tf
provider juju {}
...
```

## Additional notes

*\<Please add relevant notes & information, links to mattermost chats, other related issues/PRs, anything to help understand and QA the PR.\>*

### CODE REVIEW

**Review by hmlanigan — CHANGES_REQUESTED** [review]
This file: https://github.com/juju/terraform-provider-juju/blob/main/templates/index.md.tmpl needs updates to Authentication section related to this change. After run `make docs` to add the results.

The PR description has no QA steps.

Have you run the acceptance tests locally? Github is consistently getting failures on the lxd tests.

File: internal/provider/provider.go [hunk 1]
```diff
@@ -115,6 +125,16 @@ func (p *jujuProvider) Schema(_ context.Context, _ provider.SchemaRequest, resp
...
+			JujuClientID: schema.StringAttribute{
+				Description: fmt.Sprintf("This is the client ID to be used. This can also be set by the `%s` environment variable", JujuClientIDEnvKey),
+				Optional:    true,
+				Sensitive:   true,
+			},
+			JujuClientSecret: schema.StringAttribute{
+				Description: fmt.Sprintf("This is the client secret to be used. This can also be set by the `%s` environment variable", JujuClientSecretEnvKey),
+				Optional:    true,
+				Sensitive:   true,
+			},
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply internal/provider/provider.go]: it is not.. removed..

File: internal/provider/provider.go [hunk 2]
```diff
@@ -115,6 +125,16 @@ func (p *jujuProvider) Schema(_ context.Context, _ provider.SchemaRequest, resp
 				Optional:    true,
 				Sensitive:   true,
 			},
+			JujuClientID: schema.StringAttribute{
+				Description: fmt.Sprintf("This is the client ID to be used. This can also be set by the `%s` environment variable", JujuClientIDEnvKey),
+				Optional:    true,
+				Sensitive:   true,
+			},
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply internal/provider/provider.go]: added, ptal

File: go.mod [hunk 3]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/bflad/tfproviderlint v0.29.0
 	github.com/hashicorp/terraform-plugin-docs v0.18.0
 	// juju 3.3.0
-	github.com/juju/juju v0.0.0-20231109132148-02479f4a1bfd
+	github.com/juju/juju v0.0.0-20240406000153-4582af466744
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply go.mod:9]: understood.

File: internal/juju/applications.go [hunk 4]
```diff
@@ -1280,7 +1280,7 @@ func (c applicationsClient) computeSetCharmConfig(
 	}
 
 	apiCharmID := apiapplication.CharmID{
-		URL:    resolvedURL,
+		URL:    newURL.String(),
```
Comment (hmlanigan) [inline internal/juju/applications.go:1283]: This needs to be resolvedURL.String() instead. It's causing the acceptance test failures.

Comment (alesstimec) [reply internal/juju/applications.go]: done

**Review by cderici — APPROVED** [review]
This LGTM, just completed manual QA against `3.4` and `3.5` and it went well, everything seems to be working fine.

I was able to provide and pass through the `client_id` and the `client_secret` just fine, and also the validation against mutual exclusivity with `username/password` seems to be working fine.

Though I couldn't test it against a real JAAS setup with Juju, I'll give it a green light, as i) I confirmed that the current functionality is working without any issues, ii) all the tests are passing, and iii) I imagine @alesstimec can confirm that it's working well on jaas as he's the one who needed this change in the first place.

**Review by hmlanigan — APPROVED** [review]

**Pull request merged**